PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/StreamingLogsToolWindowManager.java

```diff
@@ -0,0 +1,107 @@
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.microsoft.intellij.helpers;
+
+import com.intellij.execution.ui.ConsoleView;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.wm.ToolWindow;
+import com.intellij.openapi.wm.ToolWindowManager;
+import com.intellij.ui.content.Content;
+import com.intellij.ui.content.ContentFactory;
+import com.intellij.ui.content.ContentManagerAdapter;
+import com.intellij.ui.content.ContentManagerEvent;
+import org.apache.commons.collections4.BidiMap;
+import org.apache.commons.collections4.bidimap.DualHashBidiMap;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class StreamingLogsToolWindowManager {
+
+    public static final String LOG_TOOL_WINDOW = "Azure Streaming Log";
+
+    private Map<Project, ToolWindow> toolWindowMap = new HashMap<>();
+    private BidiMap<String, String> resourceIdToNameMap = new DualHashBidiMap<>();
+
+    public static StreamingLogsToolWindowManager getInstance() {
+        return StreamingLogsToolWindowManager.SingletonHolder.INSTANCE;
+    }
+
+    public void showStreamingLogConsole(Project project, String resourceId, String resourceName,
+                                        ConsoleView consoleView) {
+        final ToolWindow toolWindow = getToolWindow(project);
+        ApplicationManager.getApplication().invokeLater(() -> {
+            toolWindow.show(null);
+            final ContentFactory contentFactory = ContentFactory.SERVICE.getInstance();
+            final String consoleName = getConsoleViewName(resourceId, resourceName);
+            Content content = toolWindow.getContentManager().findContent(consoleName);
+            if (content == null) {
+                content = contentFactory.createContent(consoleView.getComponent(), consoleName, false);
+                content.setDisposer(consoleView);
+                toolWindow.getContentManager().addContent(content);
+            }
+            toolWindow.getContentManager().setSelectedContent(content);
+        });
+    }
+
+    public void removeConsoleView(String resourceId) {
+        resourceIdToNameMap.remove(resourceId);
+    }
+
+    private String getConsoleViewName(String resourceId, String resourceName) {
+        if (resourceIdToNameMap.containsKey(resourceId)) {
+            return resourceIdToNameMap.get(resourceId);
+        }
+        String result = resourceName;
+        int i = 1;
+        while (resourceIdToNameMap.containsValue(result)) {
+            result = String.format("%s(%s)", resourceName, i++);
+        }
+        resourceIdToNameMap.put(resourceId, result);
+        return result;
+    }
+
+    private ToolWindow getToolWindow(Project project) {
+        if (toolWindowMap.containsKey(project)) {
```

Comment:
where is the put method for toolWindowMap?

Reply:
Added